wallet/src/main/java/com/android/identity_credential/wallet/QrEngagementViewModel.kt

```diff
@@ -0,0 +1,99 @@
+package com.android.identity_credential.wallet
+
+import android.app.Application
+import androidx.core.content.ContextCompat
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.viewModelScope
+import com.android.identity.android.mdoc.engagement.QrEngagementHelper
+import com.android.identity.android.mdoc.transport.DataTransport
+import com.android.identity.android.mdoc.transport.DataTransportOptions
+import com.android.identity.internal.Util
+import com.android.identity.mdoc.connectionmethod.ConnectionMethod
+import com.android.identity.mdoc.connectionmethod.ConnectionMethodBle
+import com.android.identity.securearea.SecureArea
+import com.android.identity.util.Logger
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.launch
+import java.security.KeyPair
+import java.util.UUID
+
+class QrEngagementViewModel(val context: Application) : AndroidViewModel(context)  {
+    companion object {
+        private const val TAG = "MainViewModel"
+    }
+
+    init {
+        Logger.i(TAG, "new mainViewModel")
+    }
+    var qrEngagementHelper: QrEngagementHelper? = null
+    private var eDeviceKeyPair: KeyPair? = null
+    private val eDeviceKeyCurve = SecureArea.EC_CURVE_P256
+
+    fun generateQrEngagementHelper(
+        onComplete: () -> Unit,
+        onConnected: () -> Unit)
+    {
+        Logger.i(TAG, "generateQrEngagementHelper called")
+        qrEngagementHelper?.close()
+        qrEngagementHelper = null
+
+        viewModelScope.launch(
+            CoroutineExceptionHandler{ _, exception ->
+                println("CoroutineExceptionHandler got $exception")
+            }
+        ) {
+            eDeviceKeyPair = Util.createEphemeralKeyPair(eDeviceKeyCurve)
+            val qrEngagementListener = object : QrEngagementHelper.Listener {
+
+                override fun onDeviceEngagementReady() {
+                    Logger.i(TAG, "QR Engagement: Device Engagement Ready")
+                }
+
+                override fun onDeviceConnecting() {
+                    Logger.i(TAG, "onDeviceConnecting")
+                }
+
+                override fun onDeviceConnected(transport: DataTransport) {
+                    Logger.i(TAG, "OnDeviceConnected via QR: qrEngagement=$qrEngagementHelper")
+
+                    onConnected()
+                    PresentationActivity.startPresentation(context, transport,
+                        qrEngagementHelper!!.handover, eDeviceKeyCurve, eDeviceKeyPair!!,
+                        qrEngagementHelper!!.deviceEngagement)
```

Comment:
So I don't think it's right to start `PresentationActivity` from here. Instead, this view model should convey to the Compose bits via state that it's in the `CONNECTED` state. Then in the Compose bits you can have can do this

```
PresentationActivity.start()
navigation.popBackStack()
```

This works because the UI will also have business logic for calling `stopQrEngagement()` when leaving the "Show QR" screen.

Reply:
I did change the PR to move `navigation.popBackStack()` out of the viewmodel, but left `PresentationActivity.startPresentation()` where it is since it depends on variables in the viewmodel (ie. eDeviceKeyPair and qrEngagementHelper)